prog/dftb+/lib_dftb/dispdftd4.F90

```diff
@@ -0,0 +1,1287 @@
+          energies(iAt2f) = energies(iAt2f) - dEr / 3.0_dp
+          energies(iAt3f) = energies(iAt3f) - dEr / 3.0_dp
+
+          gradients(:, iAt1) = gradients(:, iAt1) - dG12 - dG13
+          gradients(:, iAt2f) = gradients(:, iAt2f) + dG12 - dG23
+          gradients(:, iAt3f) = gradients(:, iAt3f) + dG13 + dG23
+
+          sigma = spread(dG12, 1, 3) * spread(vec12, 2, 3)&
+              & + spread(dG13, 1, 3) * spread(vec13, 2, 3)&
+              & + spread(dG23, 1, 3) * spread(vec23, 2, 3)
+
+          stress = stress + sigma
+
+          dc9dcn1 = (dc6dcn(iAt1, iAt2f) / c6_12 + dc6dcn(iAt1, iAt3f) / c6_13) * 0.5_dp
+          dc9dcn2 = (dc6dcn(iAt2f, iAt1) / c6_12 + dc6dcn(iAt2f, iAt3f) / c6_23) * 0.5_dp
+          dc9dcn3 = (dc6dcn(iAt3f, iAt1) / c6_13 + dc6dcn(iAt3f, iAt2f) / c6_23) * 0.5_dp
+          dEdcn(iAt1) = dEdcn(iAt1) - dc9dcn1 * dEr
+          dEdcn(iAt2f) = dEdcn(iAt2f) - dc9dcn2 * dEr
+          dEdcn(iAt3f) = dEdcn(iAt3f) - dc9dcn3 * dEr
+
+          dc9dq1 = (dc6dq(iAt1, iAt2f) / c6_12 + dc6dq(iAt1, iAt3f) / c6_13) * 0.5_dp
+          dc9dq2 = (dc6dq(iAt2f, iAt1) /c6_12 + dc6dq(iAt2f, iAt3f) / c6_23) * 0.5_dp
+          dc9dq3 = (dc6dq(iAt3f, iAt1) /c6_13 + dc6dq(iAt3f, iAt2f) /c6_23) * 0.5_dp
+          dEdq(iAt1) = dEdq(iAt1) - dc9dq1*dEr
+          dEdq(iAt2f) = dEdq(iAt2f) - dc9dq2*dEr
+          dEdq(iAt3f) = dEdq(iAt3f) - dc9dq3*dEr
+        end do
+      end do
+    end do
+
+  end subroutine threeBodyDispersionGradient
+
+
+  !> Driver for the calculation of DFT-D4 dispersion related properties.
+  subroutine dispersionEnergy(calculator, nAtom, coords, species, neigh, img2CentCell, recPoint,&
+      & energies, gradients, stress, volume, parEwald)
+
+    !> DFT-D dispersion model.
+    type(DftD4Calculator), intent(in) :: calculator
+
+    !> Nr. of atoms (without periodic images)
+    integer, intent(in) :: nAtom
+
+    !> Coordinates of the atoms (including images)
+    real(dp), intent(in) :: coords(:, :)
+
+    !> Species of every atom.
+    integer, intent(in) :: species(:)
+
+    !> Updated neighbour list.
+    type(TNeighbourList), intent(in) :: neigh
+
+    !> Mapping into the central cell.
+    integer, intent(in) :: img2CentCell(:)
+
+    !> Contains the points included in the reciprocal sum.
+    !> The set should not include the origin or inversion related points.
+    real(dp), allocatable, intent(in) :: recPoint(:, :)
+
+    !> Parameter for Ewald summation.
+    real(dp), intent(in), optional :: parEwald
+
+    !> Updated energy vector at return
+    real(dp), intent(out) :: energies(:)
+
+    !> Updated gradient vector at return
+    real(dp), intent(out) :: gradients(:, :)
+
+    !> Upgraded stress
+    real(dp), intent(out), optional :: stress(:, :)
+
+    !> Volume, if system is periodic
+    real(dp), intent(in), optional :: volume
+
+    integer, allocatable :: nNeigh(:)
+
+    real(dp) :: sigma(3, 3)
+    real(dp), parameter :: charge = 0.0_dp  ! FIXME
```

Comment:
Maybe not as finished as I thought. How do we get the molecular charge in here in a reliable way?

Reply:
Do you need the whole molecule charges or atom charges? The later is simple (but gross or net charge if so), the former would require something like a topological analysis to define 'molecules'.

Reply:
Exactly, the net charge of the entire system is needed to constraint the number of electrons for DFT-D4 correctly in the EEQ model.

Reply:
Is 'entire system' all atoms, in which case this is the net charge. Or is it net charge for each distinct molecular fragment? We can discuss details tomorrow.

Reply:
Yes, all atoms, I need the net charge. There are no molecular fragments involved here, since this would lead to a non-continuous PES.

Reply:
The total net charge of the system stays constant during the calculation (unless we have open boundary conditions as in a transport calculation). So, probably we should pass it, when `DispDftD4` becomes initialized and make sure, that D4 is disallowed for transport. On the long term, we would maybe need and `updateCharges()` call for the dispersion interface...

Reply:
If this `updateCharges()` routine is general enough to return a potential to the Hamiltonian, one might generalize the `DispIface` for self-consistent DFT-D4 and MBD with this approach.

For now setting it in the input is fine, we just have to guarantee that the net charge is set before we setup the dispersion model.